deepinv/loss/metric/metric.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import annotations
+from types import ModuleType
+from typing import Optional, TYPE_CHECKING
+
+from torch import Tensor
+from torch.nn import Module
+
+from deepinv.loss.loss import Loss
+from deepinv.loss.metric.functional import complex_abs, norm
+from deepinv.utils.plotting import rescale_img
+
+if TYPE_CHECKING:
+    from deepinv.physics.forward import Physics
+
+
+def import_pyiqa() -> ModuleType:
+    try:
+        import pyiqa
+
+        return pyiqa
+    except ImportError:
+        raise ImportError(
+            "Metric not available. Please install the pyiqa package with `pip install pyiqa`."
+        )
+
+
+class Metric(Loss):
+    r"""
+    Base class for metrics.
+
+    See docs for ``forward()`` below for more details.
+
+    To create a new metric, inherit from this class and override the function :meth:`deepinv.loss.metric.Metric.metric`.
+
+    :param bool complex_abs: perform complex magnitude before passing data to metric function. If ``True``,
+        the data must either be of complex dtype or have size 2 in the channel dimension (usually the second dimension after batch).
+    :param bool train_loss: use metric as a training loss, by returning one minus the metric. If lower is better, does nothing.
```

Comment:
I'm not sure if we should assume that metrics are always between 0 and 1, with 1 being bad.
I would remove this option.

Reply:
I've now changed it such that the train_loss option is only visible in the docs for SSIM, however I have kept it as an option for all metrics, so that it's very easy to implement for future metrics which would also require it. I have generalised it by adding a invert_metric method which can be overridden (e.g. for SSIM, it should be 1-m). I've made this clear in the docstrings.